source/slang/slang-ir-check-shader-parameter-type.cpp

```diff
@@ -0,0 +1,65 @@
+#include "slang-ir-check-shader-parameter-type.h"
+#include "slang-ir-util.h"
+
+namespace Slang
+{
+    void checkForInvalidShaderParameterTypeForMetal(IRModule* module, DiagnosticSink* sink)
+    {
+        HashSet<IRInst*> workListSet;
+        List<IRInst*> workList;
+        for (auto inst : module->getGlobalInsts())
+        {
+            if (inst->getOp() == kIROp_ParameterBlockType)
+            {
+                auto type = inst->getOperand(0);
+                if (workListSet.add(type))
+                    workList.add(type);
+                // Diagnose an error on `ParameterBlock<ConstantBuffer<T>>`.
+                if (type->getOp() == kIROp_ConstantBufferType)
+                {
+                    bool foundUseSite = false;
+                    for (auto use = inst->firstUse; use; use = use->nextUse)
+                    {
+                        auto user = use->getUser();
+                        if (user->sourceLoc.isValid())
+                        {
+                            sink->diagnose(user, Diagnostics::constantBufferInParameterBlockNotAllowedOnMetal);
+                            foundUseSite = true;
+                            break;
```

Comment:
I am a little worried that this will leads to too many duplicate errors. this is also a very unlikely corner case that probably doesn't really matter.